tests/integration_tests/functional/test_serial_io.py

```diff
@@ -0,0 +1,149 @@
+        Return True when `_input` is the same as `_string`.
+        """
+        self._input += str(input_char)
+        if self._input == self._string[:len(self._input)]:
+            if len(self._input) == len(self._string):
+                self._input = ""
+                return True
+            return False
+
+        self._input = self._input[1:]
+        return False
+
+
+class TestState(MatchStaticString):
+    """Generic test state object."""
+
+    # Prevent state objects from being collected by pytest.
+    __test__ = False
+
+    def __init__(self, match_string=''):
+        """Initialize state fields."""
+        MatchStaticString.__init__(self, match_string)
+        print('\n*** Current test state: ', str(self), end='')
+
+    def handle_input(self, microvm, input_char):
+        """Handle input event and return next state."""
+
+    def __repr__(self):
+        """Leverages the __str__ method to describe the TestState."""
+        return self.__str__()
+
+    def __str__(self):
+        """Return state name."""
+        return self.__class__.__name__
+
+
+class WaitLogin(TestState):
+    """Initial state when we wait for the login prompt."""
+
+    def handle_input(self, microvm, input_char) -> TestState:
+        """Handle input and return next state."""
+        if self.match(input_char):
+            # Send login username.
+            microvm.serial_input("root")
+            return WaitPasswordPrompt("Password:")
+        return self
+
+
+class WaitPasswordPrompt(TestState):
+    """Wait for the password prompt to be shown."""
+
+    def handle_input(self, microvm, input_char) -> TestState:
+        """Handle input and return next state."""
+        if self.match(input_char):
+            microvm.serial_input("root")
+            # Wait 1 second for shell
+            time.sleep(1)
+            microvm.serial_input("id")
+            return WaitIDResult("uid=0(root) gid=0(root) groups=0(root)")
+        return self
+
+
+class WaitIDResult(TestState):
+    """Wait for the console to show the result of the 'id' shell command."""
+
+    def handle_input(self, microvm, input_char) -> TestState:
+        """Handle input and return next state."""
+        if self.match(input_char):
+            return TestFinished()
+        return self
+
+
+class TestFinished(TestState):
+    """Test complete and successful."""
+
+    def handle_input(self, microvm, input_char) -> TestState:
+        """Return self since the test is about to end."""
+        return self
+
+
+def test_serial_console_login(test_microvm_with_ssh):
+    """Test serial console login."""
+    microvm = test_microvm_with_ssh
+    microvm.jailer.daemonize = False
+    microvm.spawn()
+
+    # We don't need to monitor the memory for this test because we are
+    # just rebooting and the process dies before pmap gets the RSS.
+    microvm.memory_events_queue = None
+
+    # Set up the microVM with 1 vCPU and a serial console.
+    microvm.basic_config(vcpu_count=1,
```

Comment:
Is there a reason why you chose to add a `boot_args` parameter to the `basic_config` function and not do it like below? I am asking becuase the `basic_config` should be a function that configures a minimal set of things (mostly things that are mandatory for the machine to boot). If any other config is needed you can choose to update that resource afterwards:
```suggestion
 microvm.basic_config(vcpu_count=1)
+
+    # Add a kernel to start booting from.
+    response = microvm.boot.put(
+        kernel_image_path=microvm.create_jailed_resource(microvm.kernel_file),
+        boot_args='console=ttyS0 reboot=k panic=1 pci=off'
+    )
+    assert microvm.api_session.is_status_no_content(response.status_code)
```

Reply:
IMO boot arguments are an important part of any basic config. With this approach you actually write less code.

Reply:
I agree with @dianpopa in the sense that `basic_config` should have default values for `mandatory` fields so it should require minimal parameters.

But that view is in fact compatible with this change since `boot_args` is a param with default `None` value, so it brings no extra complexity.

I'm voting for keeping @sandreim's change as it is flexible:
 - if you don't care about special boot args, simply call `basic_config()` like before
 - if you want boot args you can just add them in a string param instead of doing another explicit `microvm.boot.put`

Reply:
> IMO boot arguments are an important part of any basic config. With this approach you actually write less code.

I am not very strong about this, I just want to raise awareness against adding code for things that can be achieved with already existent code especially inside the CI which is already cumbersome.

I agree with all you said, `boot_args` are important for any basic config that is why they are [defaulted](https://github.com/firecracker-microvm/firecracker/blob/196c4655e679386ab6c45e6ac7579f1d2b054efd/vmm/src/lib.rs#L113) inside Firecracker. `basic_config` appeared as a solution for creating like a sample microVM to which you can attach things that are specific to your test scenario and I think we should stick to that initial purpose.